File: java/src/main/java/com/todo/ui/Menu.java
package com.todo.ui;

import com.todo.model.Task;
import com.todo.model.TaskCategory;
import com.todo.model.TaskStatus;
import com.todo.model.User;
import com.todo.service.TaskManager;
import com.todo.service.UserManager;

import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Scanner;
import java.util.UUID;

public final class Menu {

    private static final DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");

    private final UserManager userManager;
    private final TaskManager taskManager;
    private final Scanner scanner;

    private User activeUser;

    public Menu(UserManager userManager, TaskManager taskManager) {
        this.userManager = userManager;
        this.taskManager = taskManager;
        this.scanner = new Scanner(System.in);
    }

    public void start() {
        System.out.println("=============================");
        System.out.println("Collaborative To-Do List");
        System.out.println("=============================");

        boolean running = true;
        while (running) {
            try {
                if (activeUser == null && !selectUser()) {
                    running = false;
                } else {
                    running = showMainMenu();
                }
            } catch (IllegalArgumentException | NoSuchElementException exception) {
                System.out.println("Error: " + exception.getMessage());
            }
        }

        System.out.println("Application closed.");
    }

    private boolean selectUser() {
        while (activeUser == null) {
            List<User> users = userManager.getAllUsers();
            String rule = Ansi.ORANGE + "━".repeat(44) + Ansi.RESET;
            System.out.println();
            System.out.println(rule);
            System.out.println("  " + Ansi.ORANGE + Ansi.BOLD + "☕ JAVA" + Ansi.RESET
                    + Ansi.GRAY + "  ·  " + Ansi.RESET + Ansi.BOLD + "Select a user" + Ansi.RESET);
            System.out.println(rule);
            for (int index = 0; index < users.size(); index++) {
                System.out.println("   " + Ansi.CYAN + Ansi.BOLD + "[" + (index + 1) + "]" + Ansi.RESET
                        + " " + users.get(index).getName());
                Ansi.sleep(30);
            }
            System.out.println("   " + Ansi.CYAN + Ansi.BOLD + "[" + (users.size() + 1) + "]" + Ansi.RESET
                    + " Add new user");
            System.out.println("   " + Ansi.CYAN + Ansi.BOLD + "[0]" + Ansi.RESET + " Exit");
            System.out.println(rule);

            int choice = readMenuChoice(0, users.size() + 1);
            if (choice == 0) {
                return false;
            }
            Ansi.sweep(Ansi.ORANGE);
            if (choice == users.size() + 1) {
                createAndLoginUser();
            } else {
                activeUser = users.get(choice - 1);
                System.out.println("Logged in as " + activeUser.getName() + ".");
            }
        }
        return true;
    }

    private boolean showMainMenu() {
        printMainMenu();
        int choice = readMenuChoice(0, 9);
        if (choice == 0) {
            return false;
        }
        Ansi.sweep(Ansi.ORANGE);
        switch (choice) {
            case 1 -> viewAllTasks();
            case 2 -> viewMyTasks();
            case 3 -> addTask();
            case 4 -> editTask();
            case 5 -> markTaskComplete();
            case 6 -> deleteTask();
            case 7 -> filterTasks();
            case 8 -> runConcurrencyDemo();
            case 9 -> switchUser();
            default -> throw new IllegalArgumentException("Invalid menu choice.");
        }
        return true;
    }

    private void printMainMenu() {
        String[] options = {
                "[1] View all tasks",
                "[2] View my tasks",
                "[3] Add task",
                "[4] Edit task",
                "[5] Mark task complete",
                "[6] Delete task",
                "[7] Filter tasks",
                "[8] Concurrency demo",
                "[9] Switch user",
                "[0] Exit",
        };
        String rule = Ansi.ORANGE + "━".repeat(44) + Ansi.RESET;
        System.out.println();
        System.out.println(rule);
        System.out.println("  " + Ansi.ORANGE + Ansi.BOLD + "☕ JAVA" + Ansi.RESET
                + Ansi.GRAY + "  ·  " + Ansi.RESET
                + Ansi.BOLD + "Active User: " + Ansi.RESET + activeUser.getName());
        System.out.println(rule);
        for (String option : options) {
            System.out.println("   " + Ansi.CYAN + Ansi.BOLD + option.substring(0, 3) + Ansi.RESET
                    + option.substring(3));
            Ansi.sleep(30);
        }
        System.out.println(rule);
    }

    private void createAndLoginUser() {
        String name = readRequiredText("Enter user name: ");
        activeUser = userManager.createUser(name);
        System.out.println("User created. Logged in as " + activeUser.getName() + ".");
    }

    private void switchUser() {
        activeUser = null;
    }

    private void viewAllTasks() {
        printTasks(taskManager.getAllTasks());
    }

    private void viewMyTasks() {
        printTasks(taskManager.getTasksByAssignedUser(activeUser.getId()));
    }

    private void addTask() {
        String title = readRequiredText("Enter task title: ");
        String description = readOptionalText("Enter task description: ");
        TaskCategory category = readCategory();
        UUID assignedUserId = readAssignedUserId();

        Task task = taskManager.addTask(title, description, category, assignedUserId, activeUser.getId());
        System.out.println("Task added with ID: " + task.getId());
    }

    private void editTask() {
        Task task = selectTask("Enter task UUID to edit: ");
        System.out.println("Editing task: " + task.getTitle());

        String title = readRequiredText("Enter new title: ");
        String description = readOptionalText("Enter new description: ");
        TaskCategory category = readCategory();
        TaskStatus status = readStatus();
        UUID assignedUserId = readAssignedUserId();

        taskManager.editTask(task.getId(), title, description, category, status, assignedUserId);
        System.out.println("Task updated.");
    }

    private void markTaskComplete() {
        Task task = selectTask("Enter task UUID to mark complete: ");
        taskManager.markTaskComplete(task.getId());
        System.out.println("Task marked as completed.");
    }

    private void deleteTask() {
        Task task = selectTask("Enter task UUID to delete: ");
        taskManager.deleteTask(task.getId());
        System.out.println("Task deleted.");
    }

    private void filterTasks() {
        System.out.println();
        System.out.println("Filter by:");
        System.out.println("1. Category");
        System.out.println("2. Status");
        System.out.println("0. Back");

        int choice = readMenuChoice(0, 2);
        switch (choice) {
            case 1 -> filterByCategory();
            case 2 -> filterByStatus();
            case 0 -> { }
            default -> throw new IllegalArgumentException("Invalid menu choice.");
        }
    }

    private void filterByCategory() {
        TaskCategory category = readCategory();
        printTasks(taskManager.getTasksByCategory(category));
    }

    private void filterByStatus() {
        TaskStatus status = readStatus();
        printTasks(taskManager.getTasksByStatus(status));
    }

    private void runConcurrencyDemo() {
        String message = taskManager.runConcurrencyDemo(activeUser.getId());
        System.out.println(message);
    }

    private Task selectTask(String prompt) {
        if (!taskManager.hasTasks()) {
            throw new NoSuchElementException("No tasks available.");
        }
        printTasks(taskManager.getAllTasks());
        UUID taskId = readUuid(prompt);
        return taskManager.getTaskById(taskId);
    }

    private UUID readAssignedUserId() {
        displayUsers();
        UUID userId = readUuid("Enter assigned user UUID: ");
        userManager.getUserById(userId);
        return userId;
    }

    private TaskCategory readCategory() {
        TaskCategory[] categories = TaskCategory.values();
        System.out.println("Select category:");
        for (int index = 0; index < categories.length; index++) {
            System.out.println((index + 1) + ". " + categories[index]);
        }
        int choice = readMenuChoice(1, categories.length);
        return categories[choice - 1];
    }

    private TaskStatus readStatus() {
        TaskStatus[] statuses = TaskStatus.values();
        System.out.println("Select status:");
        for (int index = 0; index < statuses.length; index++) {
            System.out.println((index + 1) + ". " + statuses[index]);
        }
        int choice = readMenuChoice(1, statuses.length);
        return statuses[choice - 1];
    }

    private int readMenuChoice(int min, int max) {
        while (true) {
            System.out.print("Enter choice: ");
            String input = scanner.nextLine().trim();
            try {
                int choice = Integer.parseInt(input);
                if (choice < min || choice > max) {
                    System.out.println("Invalid menu choice. Please enter a number between " + min + " and " + max + ".");
                    continue;
                }
                return choice;
            } catch (NumberFormatException exception) {
                System.out.println("Invalid menu choice. Please enter a number.");
            }
        }
    }

    private UUID readUuid(String prompt) {
        while (true) {
            System.out.print(prompt);
            String input = scanner.nextLine().trim();
            try {
                return UUID.fromString(input);
            } catch (IllegalArgumentException exception) {
                System.out.println("Invalid UUID format. Please try again.");
            }
        }
    }

    private String readRequiredText(String prompt) {
        while (true) {
            System.out.print(prompt);
            String input = scanner.nextLine().trim();
            if (!input.isEmpty()) {
                return input;
            }
            System.out.println("Value must not be empty.");
        }
    }

    private String readOptionalText(String prompt) {
        System.out.print(prompt);
        return scanner.nextLine().trim();
    }

    private void displayUsers() {
        List<User> users = userManager.getAllUsers();
        System.out.println();
        System.out.println("Available Users");
        System.out.printf("%-36s %-20s %-16s%n", "UUID", "Name", "Created");
        System.out.println("-----------------------------------------------------------------------");
        for (User user : users) {
            System.out.printf(
                    "%-36s %-20s %-16s%n",
                    user.getId(),
                    truncate(user.getName(), 20),
                    user.getCreatedAt().format(DATE_TIME_FORMATTER)
            );
        }
    }

    private void printTasks(List<Task> tasks) {
        System.out.println();
        if (tasks.isEmpty()) {
            System.out.println("No tasks found.");
            return;
        }

        System.out.printf("%-8s %-12s %-12s %-12s %-22s %-22s%n",
                "ID", "Status", "User", "Category", "Title", "Created By");
        System.out.println("-----------------------------------------------------------------------------------------------");

        for (Task task : tasks) {
            User assignedUser = userManager.getUserById(task.getAssignedUserId());
            User createdByUser = userManager.getUserById(task.getCreatedByUserId());
            System.out.printf("%-8s %-12s %-12s %-12s %-22s %-22s%n",
                    shortId(task.getId()),
                    task.getStatus(),
                    truncate(assignedUser.getName(), 12),
                    task.getCategory(),
                    truncate(task.getTitle(), 22),
                    truncate(createdByUser.getName(), 22));
        }

        System.out.println();
        for (Task task : tasks) {
            User assignedUser = userManager.getUserById(task.getAssignedUserId());
            System.out.println("Task UUID: " + task.getId());
            System.out.println("Assigned To: " + assignedUser.getName());
            System.out.println("Description: " + (task.getDescription().isBlank() ? "(none)" : task.getDescription()));
            System.out.println("Created At: " + task.getCreatedAt().format(DATE_TIME_FORMATTER));
            System.out.println("-----------------------------------------------------------------------------------------------");
        }
    }

    private String shortId(UUID uuid) {
        return uuid.toString().substring(0, 8);
    }

    private String truncate(String value, int maxLength) {
        if (value.length() <= maxLength) {
            return value;
        }
        return value.substring(0, maxLength - 3) + "...";
    }
}

File: javascript/menu.js
const readline = require('readline');
const { CATEGORY, STATUS } = require('./constants');
const userManager = require('./userManager');
const taskManager = require('./taskManager');

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

// In piped (non-TTY) mode readline fires all 'line' events synchronously before
// the next ask() can register a listener. Buffer incoming lines and drain them
// into waiting resolvers so no input is lost regardless of mode.
const _lineQueue = [];
const _waiting   = [];

rl.on('line', line => {
  if (_waiting.length > 0) {
    _waiting.shift()(line);
  } else {
    _lineQueue.push(line);
  }
});

function ask(question) {
  process.stdout.write(question);
  return new Promise(resolve => {
    if (_lineQueue.length > 0) {
      resolve(_lineQueue.shift());
    } else {
      _waiting.push(resolve);
    }
  });
}

const IS_TTY = process.stdout.isTTY;

// Keep the startup banner visible: skip the clear on the very first screen so
// the splash stays on screen with the first menu drawn underneath it.
let _skipFirstClear = true;

const C = {
  reset: '\x1b[0m',
  bold: '\x1b[1m',
  yellow: '\x1b[38;5;226m',
  cyan: '\x1b[36m',
  gray: '\x1b[38;5;245m',
  green: '\x1b[32m',
};

const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

// Draws a colored bar running left-to-right across the screen, then clears it.
async function sweep(color = C.yellow) {
  const width = 44;
  for (let i = 1; i <= width; i++) {
    process.stdout.write('\r' + color + '━'.repeat(i) + C.reset);
    await sleep(8);
  }
  process.stdout.write('\r' + ' '.repeat(width) + '\r');
}

function hr() { console.log('─'.repeat(52)); }

function header(title) {
  if (IS_TTY && !_skipFirstClear) console.clear();
  _skipFirstClear = false;
  hr();
  console.log(`  ${title}`);
  hr();
}

function taskLine(task) {
  const icon = task.status === STATUS.COMPLETED ? '✓'
             : task.status === STATUS.IN_PROGRESS ? '◌' : '☐';
  const assignee = userManager.getUserById(task.assignedTo);
  const name = assignee ? assignee.name : 'Unassigned';
  return (
    `  [${task.id}] ${icon} ${task.title}\n` +
    `       ${task.category}  |  ${task.status}  |  → ${name}`
  );
}

async function pause() {
  await ask('\n  Press Enter to continue...');
}

// ── Screens ───────────────────────────────────────────────────────────────────

async function userSelectionScreen() {
  while (true) {
    header('Collaborative To-Do List');
    console.log('\n  ' + C.bold + C.yellow + 'Select a user' + C.reset + '\n');

    const users = userManager.getUsers();
    for (let i = 0; i < users.length; i++) {
      console.log('   ' + C.cyan + C.bold + `[${i + 1}]` + C.reset + ' ' + users[i].name);
      await sleep(30);
    }
    console.log('   ' + C.cyan + C.bold + `[${users.length + 1}]` + C.reset + ' Add new user');
    console.log('   ' + C.cyan + C.bold + '[0]' + C.reset + ' Exit\n');

    const choice = (await ask('  ➤ ')).trim();

    if (choice === '0') { shutdown(); }

    const idx = parseInt(choice) - 1;
    if (idx >= 0 && idx < users.length) {
      await sweep();
      userManager.setActiveUser(users[idx].id);
      return;
    }

    if (parseInt(choice) === users.length + 1) {
      const name = (await ask('  Name: ')).trim();
      if (name) {
        await sweep();
        const user = userManager.createUser(name);
        userManager.setActiveUser(user.id);
        return;
      }
    }
  }
}

async function viewTasksScreen(tasks) {
  if (IS_TTY) console.clear();
  header('Tasks');
  if (tasks.length === 0) {
    console.log('\n  No tasks found.');
  } else {
    console.log();
    tasks.forEach(t => console.log(taskLine(t) + '\n'));
  }
  await pause();
}

async function addTaskScreen() {
  if (IS_TTY) console.clear();
  header('Add New Task');

  const title = (await ask('\n  Title: ')).trim();
  if (!title) return;

  const description = (await ask('  Description (optional): ')).trim();

  const cats = Object.values(CATEGORY);
  console.log('\n  Category:');
  cats.forEach((c, i) => console.log(`    ${i + 1}. ${c}`));
  const catChoice = parseInt((await ask('  > ')).trim()) - 1;
  const category = cats[catChoice] ?? CATEGORY.OTHER;

  const users = userManager.getUsers();
  console.log('\n  Assign to:');
  users.forEach((u, i) => console.log(`    ${i + 1}. ${u.name}`));
  const userChoice = parseInt((await ask('  > ')).trim()) - 1;
  const assignedUser = users[userChoice] ?? userManager.getActiveUser();

  const task = await taskManager.addTask({
    title,
    description,
    category,
    assignedTo: assignedUser.id,
    createdBy: userManager.getActiveUser().id,
  });

  console.log(`\n  ✓ Task "${task.title}" added.`);
  await pause();
}

async function editTaskScreen() {
  if (IS_TTY) console.clear();
  header('Edit Task');

  const tasks = taskManager.getAllTasks();
  if (tasks.length === 0) {
    console.log('\n  No tasks to edit.');
    await pause();
    return;
  }

  console.log();
  tasks.forEach(t => console.log(taskLine(t) + '\n'));

  const id = (await ask('  Task ID to edit: ')).trim();
  const task = taskManager.getTaskById(id);
  if (!task) {
    console.log('\n  Task not found.');
    await pause();
    return;
  }

  console.log(`\n  Editing: "${task.title}" (Enter to keep current value)\n`);

  const title       = (await ask(`  Title [${task.title}]: `)).trim();
  const description = (await ask(`  Description [${task.description || 'none'}]: `)).trim();

  const statuses = Object.values(STATUS);
  console.log('\n  Status:');
  statuses.forEach((s, i) => console.log(`    ${i + 1}. ${s}`));
  const statusChoice = parseInt((await ask('  > ')).trim()) - 1;
  const status = statuses[statusChoice] ?? task.status;

  const users = userManager.getUsers();
  console.log('\n  Assign to:');
  users.forEach((u, i) => console.log(`    ${i + 1}. ${u.name}`));
  const userChoice  = parseInt((await ask('  > ')).trim()) - 1;
  const assignedUser = users[userChoice];

  await taskManager.updateTask(id, {
    ...(title       && { title }),
    ...(description && { description }),
    status,
    ...(assignedUser && { assignedTo: assignedUser.id }),
  });

  console.log('\n  ✓ Task updated.');
  await pause();
}

async function deleteTaskScreen() {
  if (IS_TTY) console.clear();
  header('Delete Task');

  const tasks = taskManager.getAllTasks();
  if (tasks.length === 0) {
    console.log('\n  No tasks to delete.');
    await pause();
    return;
  }

  console.log();
  tasks.forEach(t => console.log(taskLine(t) + '\n'));

  const id = (await ask('  Task ID to delete: ')).trim();
  const task = taskManager.getTaskById(id);
  if (!task) {
    console.log('\n  Task not found.');
    await pause();
    return;
  }

  const confirm = (await ask(`  Delete "${task.title}"? (y/n): `)).trim().toLowerCase();
  if (confirm === 'y') {
    await taskManager.deleteTask(id);
    console.log('\n  ✓ Task deleted.');
  } else {
    console.log('\n  Cancelled.');
  }
  await pause();
}

async function markCompleteScreen() {
  if (IS_TTY) console.clear();
  header('Mark Task Complete');

  const tasks = taskManager.getAllTasks().filter(t => t.status !== STATUS.COMPLETED);
  if (tasks.length === 0) {
    console.log('\n  No pending tasks.');
    await pause();
    return;
  }

  console.log();
  tasks.forEach(t => console.log(taskLine(t) + '\n'));

  const id = (await ask('  Task ID to complete: ')).trim();
  const task = taskManager.getTaskById(id);
  if (!task) {
    console.log('\n  Task not found.');
  } else {
    await taskManager.updateTask(id, { status: STATUS.COMPLETED });
    console.log(`\n  ✓ "${task.title}" marked as COMPLETED.`);
  }
  await pause();
}

async function filterScreen() {
  if (IS_TTY) console.clear();
  header('Filter Tasks');
  console.log('\n  Filter by:\n');
  console.log('    1. Category');
  console.log('    2. Status');
  console.log('    0. Back\n');

  const choice = (await ask('  > ')).trim();

  if (choice === '1') {
    const cats = Object.values(CATEGORY);
    console.log('\n  Category:');
    cats.forEach((c, i) => console.log(`    ${i + 1}. ${c}`));
    const c = parseInt((await ask('  > ')).trim()) - 1;
    if (cats[c]) await viewTasksScreen(taskManager.filterByCategory(cats[c]));
  } else if (choice === '2') {
    const statuses = Object.values(STATUS);
    console.log('\n  Status:');
    statuses.forEach((s, i) => console.log(`    ${i + 1}. ${s}`));
    const s = parseInt((await ask('  > ')).trim()) - 1;
    if (statuses[s]) await viewTasksScreen(taskManager.filterByStatus(statuses[s]));
  }
}

async function concurrencyDemoScreen() {
  if (IS_TTY) console.clear();
  header('Concurrency Demo');
  console.log('\n  Simulating concurrent task updates with async/await + mutex...\n');

  const log = await taskManager.simulateConcurrentAccess();
  log.forEach(l => console.log(l));

  console.log('\n  Promise.all fires all operations simultaneously.');
  console.log('  The AsyncMutex queues writers so each update is atomic,');
  console.log('  mirroring Java\'s synchronized blocks on a shared list.\n');
  await pause();
}

function shutdown() {
  console.log('\n  Goodbye!\n');
  rl.close();
  process.exit(0);
}

// ── Main menu loop ────────────────────────────────────────────────────────────

async function mainMenu() {
  while (true) {
    const active = userManager.getActiveUser();
    if (IS_TTY) console.clear();

    const rule = C.yellow + '━'.repeat(44) + C.reset;
    console.log('\n' + rule);
    console.log('  ' + C.yellow + C.bold + 'JS' + C.reset + C.gray + '  ·  ' + C.reset
      + C.bold + 'Logged in as: ' + C.reset + active.name);
    console.log(rule);

    const options = [
      '[1] View all tasks',
      '[2] View my tasks',
      '[3] Add task',
      '[4] Edit task',
      '[5] Mark task complete',
      '[6] Delete task',
      '[7] Filter tasks',
      '[8] Concurrency demo',
      '[9] Switch user',
      '[0] Exit',
    ];
    for (const opt of options) {
      console.log('   ' + C.cyan + C.bold + opt.slice(0, 3) + C.reset + opt.slice(3));
      await sleep(30);
    }
    console.log(rule);

    const choice = (await ask('  ➤ ')).trim();

    if (['1', '2', '3', '4', '5', '6', '7', '8', '9'].includes(choice)) {
      await sweep();
    }

    switch (choice) {
      case '1': await viewTasksScreen(taskManager.getAllTasks()); break;
      case '2': await viewTasksScreen(
        taskManager.getAllTasks().filter(t => t.assignedTo === active.id)
      ); break;
      case '3': await addTaskScreen(); break;
      case '4': await editTaskScreen(); break;
      case '5': await markCompleteScreen(); break;
      case '6': await deleteTaskScreen(); break;
      case '7': await filterScreen(); break;
      case '8': await concurrencyDemoScreen(); break;
      case '9': await userSelectionScreen(); break;
      case '0': shutdown(); break;
    }
  }
}

async function start() {
  await userSelectionScreen();
  await mainMenu();
}

module.exports = { start };
